### PR TITLE
Add support for POSIX predefined character classes

### LINKED ISSUE #4413: [FEA] Add support for POSIX characters in regular expressions
**Is your feature request related to a problem? Please describe.**

Spark regexp supports a number of POSIX character classes that we do not support on GPU (we currently fall back to CPU for any of these because we do not support `\p`).

```
POSIX character classes (US-ASCII only)
--
\p{Lower} | A lower-case alphabetic character: [a-z]
\p{Upper} | An upper-case alphabetic character:[A-Z]
\p{ASCII} | All ASCII:[\x00-\x7F]
\p{Alpha} | An alphabetic character:[\p{Lower}\p{Upper}]
\p{Digit} | A decimal digit: [0-9]
\p{Alnum} | An alphanumeric character:[\p{Alpha}\p{Digit}]
\p{Punct} | Punctuation: One of !"#$%&'()*+,-./:;<=>?@[\]^_`{\|}~
\p{Graph} | A visible character: [\p{Alnum}\p{Punct}]
\p{Print} | A printable character: [\p{Graph}\x20]
\p{Blank} | A space or a tab: [ \t]
\p{Cntrl} | A control character: [\x00-\x1F\x7F]
\p{XDigit} | A hexadecimal digit: [0-9a-fA-F]
\p{Space} | A whitespace character: [ \t\n\x0B\f\r]
```

Here is an example in Spark.

```
scala> spark.sql("SELECT s, s RLIKE '\\\\p{Lower}' FROM ss").show
+---+-----------------+
|  s|s RLIKE \p{Lower}|
+---+-----------------+
|  a|             true|
|  A|            false|
+---+-----------------+
```

**Describe the solution you'd like**
Add support for these, or document that we do not support them. If we chose to support this, we could transpile these to a form that cuDF understands.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by anthony-chang**
Closes #4413

This PR adds support for `\p{...}` for these character classes defined [here](https://docs.oracle.com/javase/8/docs/api/java/util/regex/Pattern.html) under "POSIX character classes (US-ASCII only)"

```
\p{Lower} | A lower-case alphabetic character: [a-z]
\p{Upper} | An upper-case alphabetic character:[A-Z]
\p{ASCII} | All ASCII:[\x00-\x7F]
\p{Alpha} | An alphabetic character:[\p{Lower}\p{Upper}]
\p{Digit} | A decimal digit: [0-9]
\p{Alnum} | An alphanumeric character:[\p{Alpha}\p{Digit}]
\p{Punct} | Punctuation: One of !"#$%&'()*+,-./:;<=>?@[\]^_`{\|}~
\p{Graph} | A visible character: [\p{Alnum}\p{Punct}]
\p{Print} | A printable character: [\p{Graph}\x20]
\p{Blank} | A space or a tab: [ \t]
\p{Cntrl} | A control character: [\x00-\x1F\x7F]
\p{XDigit} | A hexadecimal digit: [0-9a-fA-F]
\p{Space} | A whitespace character: [ \t\n\x0B\f\r]
```

Signed-off-by: Anthony Chang <antchang@nvidia.com>

### CODE REVIEW

**Comment by anthony-chang** [comment]
build

File: integration_tests/src/main/python/string_test.py
Comment (andygrove) [inline integration_tests/src/main/python/string_test.py:998]: Could you update the fuzz tests in `RegularExpressionTranspilerSuite` to make sure all of the characters here are included in the generated data

**Comment by andygrove** [comment]
> @anthony-chang @andygrove We are in burndown for 22.06. Is this meant to target 22.06, or should it be retargeted towards 22.08?

Good point. @anthony-chang  can you retarget for 22.08

**Review by andygrove — DISMISSED** [review]
Please retarget for 22.08

**Comment by anthony-chang** [comment]
This is blocked until the fix in #5610 is merged

**Comment by andygrove** [comment]
> This is blocked until the fix in #5610 is merged

@anthony-chang this has now been merged

**Comment by anthony-chang** [comment]
build

**Comment by anthony-chang** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
Comment (andygrove) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:433]: we should document this limitation in the compatibility guide

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala
Comment (andygrove) [inline tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:402]: Is it intentional that only a subset are tested here?

Comment (anthony-chang) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala:402]: Yes, this one is mostly a sanity check since the `\p{Print}` transpilation recursively uses the definition of the 4 `\p{}` classes above it. 
I tested the full set in the integration tests.

File: tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala
Comment (anthony-chang) [reply tests/src/test/scala/com/nvidia/spark/rapids/RegularExpressionTranspilerSuite.scala]: oops

**Review by andygrove — APPROVED** [review]

**Comment by anthony-chang** [comment]
build

**Comment by anthony-chang** [comment]
build

**Comment by anthony-chang** [comment]
I've removed `_` from the Scala fuzz tests until a fix for rapidsai/cudf#11062 is made.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
Comment (NVnavkumar) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:1206]: Any particular reason this handling is included here? Did this come about in fuzz testing?

**Review by NVnavkumar — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala
Comment (anthony-chang) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/RegexParser.scala:453]: Right, I also need to update the compatibility doc

**Comment by anthony-chang** [comment]
build

**Review by NVnavkumar — APPROVED** [review]

**Comment by anthony-chang** [comment]
build

**Pull request merged**
